lint: underef false positive in method call

If we have a type `T` that has a method `M` which receives `*T`, we shouldn't warn on this:
```go
(*v).M() // v has type T
```
Because `(*v)` makes a copy in the original code and does not in the `v.M()`.

Reproducer:
```
$GOPATH/src/github.com/coreos/etcd/etcdserver/raft.go:450:27: underef: could simplify (*cl).Member to cl.Member
```

--- a/lint/testdata/underef/negative_tests.go
+++ b/lint/testdata/underef/negative_tests.go
@@ -7,6 +7,11 @@ func sampleCase2(k sampleInterface) {
 	k.(*sampleStruct).method()
 }
 
+func methodCall(v *sampleStruct) {
+	(*v).method() // Call on copy
+	v.method()
+}
+
 func multipleIndir1() {
 	type point struct{ x, y int }
 	pt := point{}

--- a/lint/testdata/underef/positive_tests.go
+++ b/lint/testdata/underef/positive_tests.go
@@ -28,9 +28,6 @@ func sampleCase() {
 	var k *sampleStruct
 	/// could simplify (*k).field to k.field
 	(*k).field = 5
-	//TODO: could simplify (*k).method() to k.method()
-	/// could simplify (*k).method to k.method
-	(*k).method()
 	//TODO: could simplify (*k).nestedStruct.nestedField to k.nestedStruct.nestedField
 	/// could simplify (*k).nestedStruct to k.nestedStruct
 	(*k).nestedStruct.nestedField = 6

--- a/lint/underef_checker.go
+++ b/lint/underef_checker.go
@@ -13,6 +13,8 @@ func init() {
 
 type underefChecker struct {
 	checkerBase
+
+	skipRecvCopy bool
 }
 
 func (c *underefChecker) InitDocumentation(d *Documentation) {
@@ -25,6 +27,10 @@ k.field = 5
 v := a[5]`
 }
 
+func (c *underefChecker) Init() {
+	c.skipRecvCopy = c.ctx.params.Bool("skipRecvCopy", true)
+}
+
 func (c *underefChecker) VisitExpr(expr ast.Expr) {
 	switch n := expr.(type) {
 	case *ast.SelectorExpr:
@@ -32,6 +38,10 @@ func (c *underefChecker) VisitExpr(expr ast.Expr) {
 		if !ok {
 			return
 		}
+		if c.skipRecvCopy && c.isPtrRecvMethodCall(n.Sel) {
+			return
+		}
+
 		if expr, ok := expr.X.(*ast.StarExpr); ok {
 			if c.checkStarExpr(expr) {
 				c.warnSelect(n)
@@ -51,6 +61,15 @@ func (c *underefChecker) VisitExpr(expr ast.Expr) {
 			}
 		}
 	}
+}
+
+func (c *underefChecker) isPtrRecvMethodCall(fn *ast.Ident) bool {
+	typ, ok := c.ctx.typesInfo.TypeOf(fn).(*types.Signature)
+	if ok && typ != nil && typ.Recv() != nil {
+		_, ok := typ.Recv().Type().(*types.Pointer)
+		return ok && c.skipRecvCopy
+	}
+	return false
 }
 
 func (c *underefChecker) underef(x *ast.ParenExpr) ast.Expr {

